In [48]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 19: A Series of Tubes ---</h2><p>Somehow, a network packet got <span title="I know how fast it's going, but I don't know where it is.">lost</span> and ended up here.  It's trying to follow a routing diagram (your puzzle input), but it's confused about where to go.</p>
<p>Its starting point is just off the top of the diagram. Lines (drawn with <code>|</code>, <code>-</code>, and <code>+</code>) show the path it needs to take, starting by going down onto the only line connected to the top of the diagram. It needs to follow this path until it reaches the end (located somewhere within the diagram) and stop there.</p>
<p>Sometimes, the lines cross over each other; in these cases, it needs to continue going the same direction, and only turn left or right when there's no other option.  In addition, someone has left <em>letters</em> on the line; these also don't change its direction, but it can use them to keep track of where it's been. For example:</p>
<pre><code>     |          
     |  +--+    
     A  |  C    
 F---|----E|--+ 
     |  |  |  D 
     +B-+  +--+

</code></pre>

<p>Given this diagram, the packet needs to take the following path:</p>
<ul>
<li>Starting at the only line touching the top of the diagram, it must go down, pass through <code>A</code>, and continue onward to the first <code>+</code>.</li>
<li>Travel right, up, and right, passing through <code>B</code> in the process.</li>
<li>Continue down (collecting <code>C</code>), right, and up (collecting <code>D</code>).</li>
<li>Finally, go all the way left through <code>E</code> and stopping at <code>F</code>.</li>
</ul>
<p>Following the path to the end, the letters it sees on its path are <code>ABCDEF</code>.</p>
<p>The little packet looks up at you, hoping you can help it find the way.  <em>What letters will it see</em> (in the order it would see them) if it follows the path? (The routing diagram is very wide; make sure you view it without line wrapping.)</p>
</article>


In [49]:
from more_itertools import first
from tabulate import tabulate


example = """
     |          
     |  +--+    
     A  |  C    
 F---|----E|--+
     |  |  |  D 
     +B-+  +--+ 
"""


def walk(diagram: str) -> str:
    diagram = diagram.splitlines()
    if not diagram[0]:
        diagram = diagram[1:]

    rows, cols = len(diagram), len(diagram[0])
    r, c = 0, diagram[0].index("|")
    dr, dc = 1, 0
    path = []
    steps = 0
    while True:
        steps += 1
        if (
            0 <= r + dr < rows
            and 0 <= c + dc < cols
            and diagram[r][c] in "|-"
            or diagram[r][c].isalpha()
        ):
            if diagram[r][c].isalpha():
                path.append(diagram[r][c])
            if diagram[r + dr][c + dc] in "|-+" or diagram[r + dr][c + dc].isalpha():
                r, c = r + dr, c + dc
            else:
                return "".join(path), steps
        else:  # diagram[r][c] == "+":
            dr, dc = dc, dr
            if not (
                0 <= r + dr < rows
                and 0 <= c + dc < cols
                and (
                    diagram[r + dr][c + dc] in "|-+"
                    or diagram[r + dr][c + dc].isalpha()
                )
            ):
                dr, dc = -dr, -dc
            r, c = r + dr, c + dc


print(f"Example; {first(walk(example))} should be 'ABCDEF'")

Example; ABCDEF should be 'ABCDEF'


In [50]:
with open("../input/day19.txt") as f:
    puzzle = f.read()


print(f"Part I; {first(walk(puzzle))}")

Part I; GPALMJSOY


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>GPALMJSOY</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>The packet is curious how many steps it needs to go.</p>
<p>For example, using the same routing diagram from the example above...</p>
<pre><code>     |          
     |  +--+    
     A  |  C    
 F---|--|-E---+ 
     |  |  |  D 
     +B-+  +--+

</code></pre>

<p>...the packet would go:</p>
<ul>
<li><code>6</code> steps down (including the first line at the top of the diagram).</li>
<li><code>3</code> steps right.</li>
<li><code>4</code> steps up.</li>
<li><code>3</code> steps right.</li>
<li><code>4</code> steps down.</li>
<li><code>3</code> steps right.</li>
<li><code>2</code> steps up.</li>
<li><code>13</code> steps left (including the <code>F</code> it stops on).</li>
</ul>
<p>This would result in a total of <code>38</code> steps.</p>
<p><em>How many steps</em> does the packet need to go?</p>
</article>

</main>


In [51]:
from more_itertools import last


print(f"Example; {last(walk(example))} should be 38")
print(f"Part II; {last(walk(puzzle))}")

Example; 38 should be 38
Part II; 16204


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>16204</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
